### PR TITLE
Made function link_extractor

### PR DESCRIPTION
**Pull request by MrRaccoonUA**
# **What does this pull request do?**

A function is implemented that searches for links to images and changes their resolution

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -37,13 +38,45 @@ def api_search(list_of_upc):
...
+
+    df_temp = pd.DataFrame(data={})
+
+    for row in range(df.shape[0]):
+        lst = df['Images links'][row]
+        prefix = 'https://target.'
+        target_link = list(search(lst, prefix))
+        prefix = 'https://i5.'
+        link_wal = (list(search(lst, prefix)))
+        if len(target_link) > 0 and len(link_wal) == 0:
```
Comment (aartiukh) [inline main.py:68]: Please add comments to this conditions

File: main.py [hunk 2]
```diff
@@ -37,13 +38,45 @@ def api_search(list_of_upc):
...
+            size = re.sub(r"1000", "3000", target_link[0])
```
Comment (aartiukh) [inline main.py:69]: `1000` and `3000` are magic numbers, create variables for them

File: main.py (hunk 1 above)
Comment (MrRaccoonUA) [reply main.py:74]: Added

File: main.py (hunk 2 above)
Comment (MrRaccoonUA) [reply main.py]: Fixed

File: main.py [hunk 3]
```diff
@@ -37,13 +38,51 @@ def api_search(list_of_upc):
...
+    """
+
+    :param df: dataframe that has been returned from function api_search
+    :return: dataframe with a new column "Download image" which containing one link on the image
+    """
+    def search(lst, prefix):
+        l = len(prefix)
+        for item in lst:
+            if item[:l] == prefix:
+                yield item
```
Comment (aartiukh) [inline main.py:58]: 👍  for `yield`

**Review by aartiukh — CHANGES_REQUESTED** [review]

File: main.py [hunk 4]
```diff
@@ -37,13 +38,51 @@ def api_search(list_of_upc):
...
+def link_extractor(df):
+    """
+
+    :param df: dataframe that has been returned from function api_search
+    :return: dataframe with a new column "Download image" which containing one link on the image
+    """
+    def search(lst, prefix):
+        l = len(prefix)
+        for item in lst:
+            if item[:l] == prefix:
```
[the remark this answers is not in the document]
Comment (MrRaccoonUA) [reply main.py:57]: If I use `-1` here, then I will need to send all link to the function. When I use `:l` compare the beginning of each reference in the array with a prefix and if they match then I save it

File: main.py [hunk 5]
```diff
@@ -37,13 +38,51 @@ def api_search(list_of_upc):
...
     return df_upc_images
 
 
+def link_extractor(df):
+    """
+
+    :param df: dataframe that has been returned from function api_search
+    :return: dataframe with a new column "Download image" which containing one link on the image
+    """
+    def search(lst, prefix):
```
[the remark this answers is not in the document]
Comment (MrRaccoonUA) [reply main.py:54]: Fixed

File: main.py [hunk 6]
```diff
@@ -37,13 +38,51 @@ def api_search(list_of_upc):
...
+            if item[:l] == prefix:
+                yield item
+
+    df_temp = pd.DataFrame(data={})
+    """
+    prefix: Start of the link which we need to find in the list of the links and extract all link with this start.
+    https://target. - start of the Target link
+    https://i5. - start of the Walmart link
+    """
+    target_default_size = 1000
```
[the remark this answers is not in the document]
Comment (MrRaccoonUA) [reply main.py:66]: Fixed

File: main.py [hunk 7]
```diff
@@ -37,13 +38,51 @@ def api_search(list_of_upc):
...
+    target_max_size = 3000
+    for row in range(df.shape[0]):
+        lst = df['Images links'][row]
+        prefix = 'https://target.'
+        target_link = list(search(lst, prefix))
+        prefix = 'https://i5.'
+        link_wal = (list(search(lst, prefix)))
```
[the remark this answers is not in the document]
Comment (MrRaccoonUA) [reply main.py:73]: Fixed

File: main.py (hunk 3 above)
Comment (MrRaccoonUA) [reply main.py:58]: Thank you

File: main.py [hunk 8]
```diff
@@ -37,13 +38,51 @@ def api_search(list_of_upc):
...
+    :param df: dataframe that has been returned from function api_search
+    :return: dataframe with a new column "Download image" which containing one link on the image
+    """
+    def extractor(lst, prefix):
+        l = len(prefix)
+        for item in lst:
+            if item[:l] == prefix:
+                yield item
+
+    df_temp = pd.DataFrame(data={})
```
Comment (aartiukh) [inline main.py:60]: Please give `df_temp` more meaningful name

File: main.py [hunk 9]
```diff
@@ -37,13 +38,51 @@ def api_search(list_of_upc):
...
 
 
+def link_extractor(df):
+    """
+
+    :param df: dataframe that has been returned from function api_search
+    :return: dataframe with a new column "Download image" which containing one link on the image
+    """
+    def extractor(lst, prefix):
+        l = len(prefix)
```
Comment (aartiukh) [inline main.py:55]: Drop this var, do not use `l` as a var name, it looks like `1`

File: main.py [hunk 10]
```diff
@@ -37,13 +38,51 @@ def api_search(list_of_upc):
...
+    df_temp = pd.DataFrame(data={})
+    """
+    prefix: Start of the link which we need to find in the list of the links and extract all link with this start.
+    https://target. - start of the Target link
+    https://i5. - start of the Walmart link
+    """
+    TARGET_DEFAULT_SIZE = 1000
+    TARGET_MAX_SIZE = 3000
+    for row in range(df.shape[0]):
+        lst = df['Images links'][row]
```
Comment (aartiukh) [inline main.py:69]: Rename `lst`

File: main.py [hunk 11]
```diff
@@ -37,13 +38,51 @@ def api_search(list_of_upc):
...
+        prefix = 'https://target.'
+        target_link = list(extractor(lst, prefix))
+        prefix = 'https://i5.'
+        link_wal = list(extractor(lst, prefix))
+        if len(target_link) > 0 and len(link_wal) == 0:
```
Comment (aartiukh) [inline main.py:74]: Get rid of these ifs. Try to get target imegas first, then check if len > 0 and return df on success. On failure, try to get walmart image and again check if len > 0 ...

**Review by aartiukh — CHANGES_REQUESTED** [review]

File: main.py (hunk 11 above)
Comment (MrRaccoonUA) [reply main.py]: Fixed in the Refactoring

File: main.py (hunk 10 above)
Comment (MrRaccoonUA) [reply main.py]: Fixed in the Refactoring

File: main.py (hunk 9 above)
Comment (MrRaccoonUA) [reply main.py]: Fixed in the Refactoring

File: main.py (hunk 8 above)
Comment (MrRaccoonUA) [reply main.py]: Fixed in the Refactoring

File: main.py [hunk 12]
```diff
@@ -1,58 +1,86 @@
...
-            for item in data['items']:
-                df_upc_images = df_upc_images.append({'UPC': item['ean'], 'Product Name': item['title'],
-                                                     'Images links': item['images']}, ignore_index=True)
-        logger.info(f'{list_of_upc.index(code)} UPC: {code} - Status: {data["code"]}. '
-                    f'Iteration time: {datetime.now() - start_time}')
-        df_upc_images['UPC'] = df_upc_images['UPC'].apply(int)
-    return df_upc_images
+    resp = requests.get('https://api.upcitemdb.com/prod/trial/lookup?upc=' + code, headers=headers)
+    data = json.loads(resp.text)
+    detail_doc = dict()
```
Comment (aartiukh) [inline main.py:22]: @MrRaccoonUA  Use named tuple instead of dict.
See https://www.accelebrate.com/blog/named-tuples-in-python for the reference

Comment (MrRaccoonUA) [reply main.py]: Changed

File: main.py [hunk 13]
```diff
@@ -1,58 +1,88 @@
...
+    resp = requests.get('https://api.upcitemdb.com/prod/trial/lookup?upc=' + code, headers=headers)
+    data = json.loads(resp.text)
+    logger.info(f'UPC: {code} - Status: {data["code"]}')
+    if data['code'] == 'OK':
+        product_info = namedtuple('Product', 'UPC Product_Name Images_Links')
+        for item in data['items']:
+            product_info = product_info(item['ean'], item['title'], item['images'])
+        return product_info
+    else:
+        return None
```
Comment (aartiukh) [inline main.py:30]:
```suggestion
    if data['code'] != 'OK':
        return None
 
    product_info = namedtuple('Product', 'UPC Product_Name Images_Links')
    for item in data['items']:
        product_info = product_info(item['ean'], item['title'], item['images'])
    return product_info
        
```

File: main.py [hunk 14]
```diff
@@ -1,58 +1,88 @@
...
+    TARGET_DEFAULT_SIZE = 1000
+    TARGET_MAX_SIZE = 3000
+    prefix = 'https://target.'
+    target_prefix = list(extractor(list_of_links, prefix))
+    if target_prefix:
+        one_link = namedtuple('New_Link', 'resized_link')
+        resized_link = re.sub(str(TARGET_DEFAULT_SIZE), str(TARGET_MAX_SIZE), target_prefix[0])
+        one_link = one_link(resized_link)
+        logger.info(f'UPC: {upc} - Status: Image OK')
+        return one_link
```
Comment (aartiukh) [inline main.py:54]: add empty line

File: main.py [hunk 15]
```diff
@@ -1,58 +1,88 @@
...
+    """
+    prefix: Start of the link which we need to find in the list of the links and extract all link with this start.
+    https://target. - start of the Target link
+    https://i5. - start of the Walmart link
+    """
+    TARGET_DEFAULT_SIZE = 1000
+    TARGET_MAX_SIZE = 3000
+    prefix = 'https://target.'
+    target_prefix = list(extractor(list_of_links, prefix))
+    if target_prefix:
```
Comment (aartiukh) [inline main.py:49]: 2 `if`s -- two functions

File: main.py [hunk 16]
```diff
@@ -1,58 +1,88 @@
...
-        time.sleep(API_COOL_DOWN_TIMEOUT)
-        if data['code'] == 'OK':
-            for item in data['items']:
-                df_upc_images = df_upc_images.append({'UPC': item['ean'], 'Product Name': item['title'],
-                                                     'Images links': item['images']}, ignore_index=True)
-        logger.info(f'{list_of_upc.index(code)} UPC: {code} - Status: {data["code"]}. '
-                    f'Iteration time: {datetime.now() - start_time}')
-        df_upc_images['UPC'] = df_upc_images['UPC'].apply(int)
-    return df_upc_images
+    resp = requests.get('https://api.upcitemdb.com/prod/trial/lookup?upc=' + code, headers=headers)
```
Comment (aartiukh) [inline main.py:21]: Create a vat for this `'https://api.upcitemdb.com/prod/trial/lookup?upc='`

File: main.py [hunk 17]
```diff
@@ -1,58 +1,88 @@
...
+    data = json.loads(resp.text)
+    logger.info(f'UPC: {code} - Status: {data["code"]}')
+    if data['code'] == 'OK':
+        product_info = namedtuple('Product', 'UPC Product_Name Images_Links')
+        for item in data['items']:
+            product_info = product_info(item['ean'], item['title'], item['images'])
```
Comment (aartiukh) [inline main.py:27]: Create a list for image links and then create a tuple

Comment (MrRaccoonUA) [reply main.py]: Done

File: main.py (hunk 16 above)
Comment (MrRaccoonUA) [reply main.py]: Done

File: main.py (hunk 15 above)
Comment (MrRaccoonUA) [reply main.py]: Done

File: main.py (hunk 14 above)
Comment (MrRaccoonUA) [reply main.py:49]: Done

File: main.py (hunk 13 above)
Comment (MrRaccoonUA) [reply main.py]: Done

**Review by aartiukh — APPROVED** [review]

**Pull request merged**